File: generated-code/codecraft/java/src/main/java/model/EntityType.java
package model;

import util.StreamUtil;

/**
 * Entity type
 */
public enum EntityType {
    /**
     * Wall, can be used to prevent enemy from moving through
     */
    WALL(0),
    /**
     * House, used to increase population
     */
    HOUSE(1),
    /**
     * Base for recruiting new builder units
     */
    BUILDER_BASE(2),
    /**
     * Builder unit can build buildings
     */
    BUILDER_UNIT(3),
    /**
     * Base for recruiting new melee units
     */
    MELEE_BASE(4),
    /**
     * Melee unit
     */
    MELEE_UNIT(5),
    /**
     * Base for recruiting new ranged units
     */
    RANGED_BASE(6),
    /**
     * Ranged unit
     */
    RANGED_UNIT(7),
    /**
     * Resource can be harvested
     */
    RESOURCE(8),
    /**
     * Ranged attacking building
     */
    TURRET(9);

    public int tag;

    EntityType(int tag) {
        this.tag = tag;
    }

    /**
     * Read EntityType from input stream
     */
    public static EntityType readFrom(java.io.InputStream stream) throws java.io.IOException {
        switch (StreamUtil.readInt(stream)) {
        case 0:
            return WALL;
        case 1:
            return HOUSE;
        case 2:
            return BUILDER_BASE;
        case 3:
            return BUILDER_UNIT;
        case 4:
            return MELEE_BASE;
        case 5:
            return MELEE_UNIT;
        case 6:
            return RANGED_BASE;
        case 7:
            return RANGED_UNIT;
        case 8:
            return RESOURCE;
        case 9:
            return TURRET;
        default:
            throw new java.io.IOException("Unexpected tag value");
        }
    }
}

File: generated-code/codecraft/kotlin/src/main/kotlin/model/EntityType.kt
package model

import util.StreamUtil

enum class EntityType private constructor(val tag: Int) {
    WALL(0),
    HOUSE(1),
    BUILDER_BASE(2),
    BUILDER_UNIT(3),
    MELEE_BASE(4),
    MELEE_UNIT(5),
    RANGED_BASE(6),
    RANGED_UNIT(7),
    RESOURCE(8),
    TURRET(9);

    companion object {
        @Throws(java.io.IOException::class)
        fun readFrom(stream: java.io.InputStream): EntityType {
            return when (StreamUtil.readInt(stream)) {
            WALL.tag -> WALL
            HOUSE.tag -> HOUSE
            BUILDER_BASE.tag -> BUILDER_BASE
            BUILDER_UNIT.tag -> BUILDER_UNIT
            MELEE_BASE.tag -> MELEE_BASE
            MELEE_UNIT.tag -> MELEE_UNIT
            RANGED_BASE.tag -> RANGED_BASE
            RANGED_UNIT.tag -> RANGED_UNIT
            RESOURCE.tag -> RESOURCE
            TURRET.tag -> TURRET
            else -> throw java.io.IOException("Unexpected tag value")
            }
        }
    }
}